openmc/filter.py

```diff
@@ -456,74 +455,97 @@ class IntegralFilter(Filter):
 
-    def check_bins(self, bins):
-        cv.check_iterable_type('filter bins', bins, Integral)
+        # Check the bin values.
+        cv.check_iterable_type('filter bins', bins, (Integral, openmc.Universe))
         for edge in bins:
-            cv.check_greater_than('filter bin', edge, 0, equality=True)
+            if isinstance(edge, Integral):
+                cv.check_greater_than('filter bin', edge, 0, equality=True)
+
+        # Extract id values.
+        if any([isinstance(b, openmc.Universe) for b in bins]):
+            bins = np.atleast_1d([b.id if isinstance(b, openmc.Universe) else b
+                                  for b in bins])
+
+        self._bins = bins
+
+    @num_bins.setter
+    def num_bins(self, num_bins): pass
 
 
-class UniverseFilter(IntegralFilter):
-    """Bins tally event locations based on the universe they occured in.
+class MaterialFilter(Filter):
+    """Bins tally event locations based on the Material they occured in.
 
     Parameters
     ----------
-    bins : Integral or Iterable of Integral
-        openmc.Universe IDs.
+    bins : openmc.Material, Integral, or iterable thereof
+        The Materials to tally.  Either openmc.Material objects or their
+        Integral ID numbers can be used.
 
     Attributes
     ----------
-    bins : Integral or Iterable of Integral
-        openmc.Universe IDs.
+    bins : Iterable of Integral
+        openmc.Material IDs.
     num_bins : Integral
         The number of filter bins
     stride : Integral
         The number of filter, nuclide and score bins within each of this
         filter's bins.
 
     """
+    @property
+    def bins(self):
+        return self._bins
 
+    @property
+    def num_bins(self):
+        return len(self.bins)
 
-class MaterialFilter(IntegralFilter):
-    """Bins tally events based on which material they occured in.
+    @bins.setter
+    def bins(self, bins):
+        # Format the bins as a 1D numpy array.
+        bins = np.atleast_1d(bins)
 
-    Parameters
-    ----------
-    bins : Integral or Iterable of Integral
-        openmc.Material IDs.
+        # Check the bin values.
+        cv.check_iterable_type('filter bins', bins, (Integral, openmc.Material))
+        for edge in bins:
+            if isinstance(edge, Integral):
+                cv.check_greater_than('filter bin', edge, 0, equality=True)
 
-    Attributes
-    ----------
-    bins : Integral or Iterable of Integral
-        openmc.Material IDs.
-    num_bins : Integral
-        The number of filter bins
-    stride : Integral
-        The number of filter, nuclide and score bins within each of this
-        filter's bins.
+        # Extract id values.
+        if any([isinstance(b, openmc.Material) for b in bins]):
+            bins = np.atleast_1d([b.id if isinstance(b, openmc.Material) else b
+                                  for b in bins])
 
-    """
+        self._bins = bins
 
+    @num_bins.setter
+    def num_bins(self, num_bins): pass
```

Comment:
If you have an `@property; def num_bins` you also need the `@num_bins.setter`.  But I don't actually want this setter to do anything because I'd prefer `.num_bins` be consistent with the `.bins` attribute.  We also don't want to raise an error though because that will be annoying when we import form HDF5 (where we need to specify `num_bins` for some filters that are not so obvious).  Hence the `pass`.  I added a comment to explain this in the source code.